### PR TITLE
Fetch readme from source archive if available for crate details page

### LINKED ISSUE #1866: Is it possible to increase a README character limit?
### Crate name

tabled

### Build failure link

https://docs.rs/crate/tabled/0.9.0

### Requested RAM limit

_No response_

### Requested timeout

_No response_

### Requested number of targets

_No response_

### Why your crate needs the resource increases

Hi there,

Yes....
I guess someone need to redo the README but is it possible to increase the limit?

Take care.

### PR DESCRIPTION
**Pull request by Nemo157**
Fixes #1866 and will allow to stop storing the readme in the database for new builds

### CODE REVIEW

File: src/web/crate_details.rs [hunk 1]
```diff
@@ -91,6 +92,34 @@ pub struct Release {
...
+    let manifest = storage.fetch_source_file(name, version, "Cargo.toml", archive_storage)?;
+    let manifest = String::from_utf8(manifest.content)
+        .context("parsing Cargo.toml")?
+        .parse::<toml::Value>()
+        .context("parsing Cargo.toml")?;
+    let paths = match manifest.get("package").and_then(|p| p.get("readme")) {
+        Some(toml::Value::Boolean(true)) => vec!["README.md"],
+        Some(toml::Value::Boolean(false)) => vec![],
+        Some(toml::Value::String(path)) => vec![path.as_ref()],
+        _ => vec!["README.md", "README.txt", "README"],
```
Comment (syphar) [inline src/web/crate_details.rs:111]: isn't this logic that we would also find somewhere in the `cargo` codebase? or some shared library?

File: src/web/crate_details.rs [hunk 2]
```diff
@@ -91,6 +92,34 @@ pub struct Release {
...
+    };
+    for path in &paths {
+        if let Ok(readme) = storage.fetch_source_file(name, version, path, archive_storage) {
```
Comment (syphar) [inline src/web/crate_details.rs:114]: we should only skip over the errors when it's an `PathNotFoundError`. Other cases should fail the function, should they?

File: src/web/crate_details.rs [hunk 3]
```diff
@@ -237,6 +266,13 @@ impl CrateDetails {
         Ok(Some(crate_details))
     }
 
+    fn enrich_readme(&mut self, storage: &Storage) -> Result<()> {
```
Comment (syphar) [inline src/web/crate_details.rs:269]: nit: shouldn't this be just `fetch_readme`, returning the readme?

File: src/web/crate_details.rs [hunk 4]
```diff
@@ -352,13 +389,19 @@ pub(crate) async fn crate_details_handler(
...
             &mut *conn,
             &params.name,
             &version,
             &version_or_latest,
             Some(&repository_stats_updater),
-        )
+        )?;
+        if let Some(ref mut details) = details {
+            if let Err(e) = details.enrich_readme(&storage) {
+                tracing::debug!("{e:?}")
```
Comment (syphar) [inline src/web/crate_details.rs:401]: what are the errors you are envisioning here? 

Why not just fetch the readme and return errors? 

Or is this just a (kind of) slow rollout?

**Review by syphar — COMMENTED** [review]
awesome improvement! thank you for working on this!

IMO the performance change is small-ish, since 
- file-exist queries are executed on the local archive index, so at most 2 S3 requests are executed. ( 3 when we start cleaning local indexes)
- the page is cached in the CDN anyways 

I had some small comments & questions.

File: src/web/crate_details.rs (hunk 4 above)
Comment (Nemo157) [reply src/web/crate_details.rs:401]: The main one I see happening is the source archive being missing, I believe there was a period where we didn't save it for some kinds of build failures. Other than that, I don't want to fail if theres other things I've missed.

What I should probably do is used debug for just the cases I expect and report other errors to sentry so we can fix them.

File: src/web/crate_details.rs (hunk 1 above)
Comment (Nemo157) [reply src/web/crate_details.rs:111]: https://github.com/rust-lang/cargo/blob/9a7c1fe322a9c4494ffd2c2684626964d1b06c83/src/cargo/util/toml/mod.rs#L2992

`cargo` is a massive library dependency we don't want to pull in though, I don't think there's any smaller libraries implementing this part of it.

File: src/web/crate_details.rs (hunk 3 above)
Comment (Nemo157) [reply src/web/crate_details.rs]: Merged this up into the caller, and moved the `fetch_readme_from_storage` into an associated method.

File: src/web/crate_details.rs (hunk 4 above)
Comment (Nemo157) [reply src/web/crate_details.rs]: Changed this to only ignore the readme being missing, all other errors will get reported to sentry.

File: src/web/crate_details.rs (hunk 2 above)
Comment (Nemo157) [reply src/web/crate_details.rs]: Yep, changed to do that.

**Pull request merged**